From aa8fe94378a14ea8192f2ea6fcd9e27856a7deec Mon Sep 17 00:00:00 2001
From: Adham Abozaeid <adham.abozaeid@microchip.com>
Date: Mon, 16 Sep 2019 10:57:07 -0700
Subject: [PATCH 12/17] staging: wilc1000: look for rtc_clk clock

If rtc_clk is provided from DT, use it and enable it.
This is optional.
The signal may be hardcoded and no need to be requested,
but if DT provides it, use it.

Signed-off-by: Eugen Hristev <eugen.hristev@microchip.com>
Acked-by: Ajay Singh <ajay.kathat@microchip.com>
Link: https://lore.kernel.org/r/1568037993-4646-2-git-send-email-eugen.hristev@microchip.com
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Adham Abozaeid <adham.abozaeid@microchip.com>
---
 drivers/staging/wilc/wilc_sdio.c          | 12 +++++++++++-
 drivers/staging/wilc/wilc_wfi_netdevice.h |  1 +
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/wilc/wilc_sdio.c b/drivers/staging/wilc/wilc_sdio.c
index 16c25d62dc63..d16ec8b24ce6 100644
--- a/drivers/staging/wilc/wilc_sdio.c
+++ b/drivers/staging/wilc/wilc_sdio.c
@@ -4,6 +4,7 @@
  * All rights reserved.
  */
 
+#include <linux/clk.h>
 #include <linux/mmc/sdio_func.h>
 #include <linux/mmc/host.h>
 #include <linux/mmc/card.h>
@@ -136,7 +137,7 @@ static int wilc_sdio_cmd53(struct wilc *wilc, struct sdio_cmd53 *cmd)
 }
 
 static int wilc_sdio_probe(struct sdio_func *func,
-			    const struct sdio_device_id *id)
+			   const struct sdio_device_id *id)
 {
 	struct wilc *wilc;
 	int ret, io_type;
@@ -164,6 +165,12 @@ static int wilc_sdio_probe(struct sdio_func *func,
 	wilc->dt_dev = &func->card->dev;
 	sdio_priv->wl = wilc;
 
+	wilc->rtc_clk = devm_clk_get(&func->card->dev, "rtc_clk");
+	if (PTR_ERR_OR_ZERO(wilc->rtc_clk) == -EPROBE_DEFER)
+		return -EPROBE_DEFER;
+	else if (!IS_ERR(wilc->rtc_clk))
+		clk_prepare_enable(wilc->rtc_clk);
+
 	if (!init_power) {
 		wilc_wlan_power_on_sequence(wilc);
 		init_power = 1;
@@ -179,6 +186,9 @@ static void wilc_sdio_remove(struct sdio_func *func)
 {
 	struct wilc *wilc = sdio_get_drvdata(func);
 
+	if (!IS_ERR(wilc->rtc_clk))
+		clk_disable_unprepare(wilc->rtc_clk);
+
 	wilc_netdev_cleanup(wilc);
 	wilc_bt_deinit();
 }
diff --git a/drivers/staging/wilc/wilc_wfi_netdevice.h b/drivers/staging/wilc/wilc_wfi_netdevice.h
index 3c8b4e8fdc46..0cbc45f2fbca 100644
--- a/drivers/staging/wilc/wilc_wfi_netdevice.h
+++ b/drivers/staging/wilc/wilc_wfi_netdevice.h
@@ -252,6 +252,7 @@ struct wilc {
 #else
 	int gpio_irq;
 #endif
+	struct clk *rtc_clk;
 	bool initialized;
 	int dev_irq_num;
 	int close;
-- 
2.23.0

